feat: refactor ActiveSession UI

mipro98 opened the pull request:
This PR is the effort to simplify and improve ActiveSession visually and also logically in the UI domain.

## Todo

- [x] fix recording Service Notification
- [x] better Pause UI
- [x] tweak Metronome layout slightly for better aesthetics
- [x] move one-time-events to Channels instead of UI States
- [x] integrate Recorder back in activeSession
   - [x] Playback Waveform inside of list element
   - [x] playback duration under Waveform
   - [ ] <del>handle recording during playback and vice versa</del> -> Not so important atm
   - [x] handle BottomSheet size change when switching Metronome <-> Recorder
   - [x] remove viewpager in BottomSheet?
   - [x] fix previews for recorder
   - [x] delete recordings (swipe2delete implemented, actual deletion not yet)
   - [x] optimize recomposition efficiency -> new issue
   - [x] fix width on small devices / landscape
   - [x] fix weird bug where bottomsheet content disappears while pager is scrolling -> new issue
- [x] redesign "New Item" screen
- [x] Cleanup legacy code
- [ ] <del>Add indicator / index number or start time of elements (so that user uderstands that most recent is at the top)</del> postpone for now
- [x] fix recomposition issues (!!)
- [ ] <del>change display style of time in SectionsList</del> postpone for now
- [x] fix dismissContainer leaking its background color beneath the section element on rounded edges
- [x] Create Compose Previews for all UI elements.
- [x] Landscape Layout 
- [x] Tablet Layout
- [x] finalize New Item screen
   - [x] <del>item/folder creation possibility</del> -> issue
   - [x] folders only shown if at least 1 folder
   - [x] fix sizing of components
   - [x] evaluate top tabs for folder selection
   - [x] fix "no folder" elements not shown
   - [ ] <del>>50% height when partially expanded</del> it's ok for now
   - [x] fix statusbar window inset or prevent 100% if not possible
   - [x] fix last practiced display + currently running item
- [x] <del>fix library colors compatibility with dark mode</del>
- [x] <del>adapt + add missing colors of colorScheme to MusikusColorScheme</del>
- [x] undo delete section
- [x] handle actions while paused
    - [x] new Item
    - [x] section deleted


----

* fixes #65 
* fixes #23 
* fixes #53 
* fixes #21 
* fixes #64

mipro98 commented:
Currently, it looks like this:

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/906b38cd-b8d9-4715-8ac8-20c7bef2442c)


However, from a color standpoint, I would rather have it to look more like this (darker background than containers). However, as one can see on the right side, it breaks the dark theme if I invert the colors with `surfaceContainer[Low(est)|High(est)]`:

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/7f5cf52f-30df-4664-913e-675da11694d0)


Apparently, this seems to be somehow possible because it is explained [here](https://m3.material.io/styles/color/roles). If you look at the [picture](https://firebasestorage.googleapis.com/v0/b/design-spec/o/projects%2Fgm3sandbox%2Fimages%2Flmjmkt4k-%5B1P%5D-color-surface-mapping-surface-complex.gif?alt=media&token=975addc3-1a35-4b94-a443-228bcf54b73c), it cycles between different "modes" of hierachies where there's also one where higher surfaces contain a *brighter* background color. Yet, I don't know how to achieve this.

matthiasemde commented:
>   * [ ]  or overlap the FAB with bottomSheet just like [here](https://medium.com/android-bits/android-anchoring-views-to-bottom-sheet-9c9069caf7d4)

This!! With a centered, expanded FAB 😍

mipro98 commented:
> > * [ ]   or overlap the FAB with bottomSheet just like [here](https://medium.com/android-bits/android-anchoring-views-to-bottom-sheet-9c9069caf7d4)
> 
> This!! With a centered, expanded FAB 😍

Centered expanded FAB will force us to use an additional padding at the top, when right-aligned we could get away without the extra space (though we cannot use the extended FAB :()

On the other side - since we don't have a drag handle - I thought about adding a small "extenend" (or "x" when extended) icon button next to the tabs in the right top corner which would interfere with the FAB...

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/f3aad257-bc09-4ad2-a836-5ac16e6cda23)

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+// TODO UIText for all Strings?
 
+@Stable
+enum class SessionPausedResumedState {
```
I think we can either call this class `ActiveSessionState` in case we want to extend it later or simply remove it and use a Boolean `isSessionRunning` instead.

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+    PAUSED
+}
+
+/** Specific one-time events triggering some UI action in ActiveSession */
```
Let's talk about how we want to send one-time events to the active session. You can watch [this](https://youtu.be/njchj9d_Lf8?si=7Y4qFeGC_qy1TrPU) Video by Philip

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+/** Specific one-time events triggering some UI action in ActiveSession */
+@Stable
+data class ActiveSessionEventStates (
+    val sessionSaved: Boolean = false,
```
We should also talk about whether we want default values for UiStates or not and then make it consistent across the app.

mipro98 replied:
I think default values are quite useful. Mainly, I set them all while doing the previews to not having to pass all parameters to every Composable.

Having default values is also a common practice amongst many guides I found (e.g. [here](https://medium.com/@MrAndroid/android-architecture-with-mvvm-and-uistate-f29aa5494465) or [here](https://craigrussell.io/2017/11/30/representing-view-state-with-kotlin-data-classes)) and even [Google does it](https://github.com/google-developer-training/basic-android-kotlin-compose-training-unscramble/blob/viewmodel/app/src/main/java/com/example/unscramble/ui/GameUiState.kt)

I see the following problems when not having default values:
* Everytime a StateFlow is created, in the `stateIn`, all parameters must be passed as default. This will result in an enourmous augmentation of boilerplate code which makes it even more convoluted than it already is (because of `stateIn` boilerplate) without adding any functionality. This will make developing experience much worse IMO
* For previews, it's really cumbersome having to pass all parameters to every UI state and slows down creation of previews

Plus it comes with the benefit of having all default values easily visible in one place along the class definition itself.

matthiasemde replied:
Yes, the boilerplate code is annoying. However, the "assembly" of the UiState doesn't/shouldn't contain much or any logic and thus does not need to be maintained so much. The `combineStateFlow` wrapper seems like a good idea.

Not using default values does have some advantages in my mind:
- Developers are forced to assign values to each property inside the `combine` function.
- Most UiStates are instantiated only once (twice if you count the `stateIn` separately). Because of that default all the values for a UiState are in one place, while the class declaration focuses solely on the **structure** of the UiState.
- The default values used for the UiState are mostly empty strings and lists which are not useful for previews anyways, right?

That's all I can think of for the moment. Most importantly I would like to keep it consistent within the app so maybe we change it for this PR?

mipro98 replied:
The big issue was the Previews. E.g. if you want to create a preview of the main screen you would also have to provide dummy data for every dialog ui state or secondary ui state you don't show in your preview. But since we may change to nullable States for non-permanently visible elements anyways we can think about it.

Also, the stateIn default arg clutter was also a big reason. Maybe we can circumvent it with the combine wrapper (although it does not appear to be universally adaptable to any number of input flows) or we can define the default UI states seperately and just pass the default variable (the default of `stateIn` actually is in 99,9% irrellevant anyways except there is an animation or so which delays distributing new values...)

matthiasemde replied:
> Am I wrong or isn't the "assembly" where most/all of the ui logic is implemented?
You are not wrong, assembling the UiState does require some logic. What I meant is that any major data processing logic should happen inside a _use case_ and what you call 'ui logic' is more a kind of 'extraction' of Integers and Strings from the data provided by use cases, _ViewModelState_ and combinations thereof.

> The big issue was the Previews. E.g. if you want to create a preview of the main screen you would also have to provide dummy data for every dialog ui state or secondary ui state you don't show in your preview. But since we may change to nullable States for non-permanently visible elements anyways we can think about it.
Ok, now it makes sense why you would definitely need  default  values. However, if we set the convention, that all UiStates that describe a UI element which is not shown are null, then we don't have that problem anymore, do we?

> Also, the stateIn default arg clutter was also a big reason. Maybe we can circumvent it with the combine wrapper (although it does not appear to be universally adaptable to any number of input flows) or we can define the default UI states seperately and just pass the default variable (the default of `stateIn` actually is in 99,9% irrellevant anyways except there is an animation or so which delays distributing new values...)
I totally agree it's annoying... Maybe we can just collapse the `stateIn` in Android Studio?

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+@Stable
+data class ActiveSessionTimerUiState(
+    val timerText: String = "00:00", // Prevents size change animation to trigger when opening the screen
+    val subHeadingAppearance: SessionPausedResumedState = SessionPausedResumedState.RUNNING,
```
Imo the type and name of a property should match better. I'd call the variable `isSessionPaused` and make it a Boolean.

mipro98 replied:
I wanted to define the UI state as "close" as possible to the actual description of *how it looks* not *what type of data it gets and interprets it into a specific appearance*. It's a minor thing, but I tried to define UIState variables as much as possible of a description of "how it looks" and not *why it looks as it looks*.

It's a question of taste I guess, also not to say that `subHeadingAppearance` is an especially well name for it.

matthiasemde replied:
Well in this specific case, the property name refers to _how it looks_ but the type still refers to _why it looks the way it does_. That's the inconsistency which irritates me I think 🙃

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+)
+
+@Stable
+data class ActiveSessionCurrentItemUiState(
```
If we simply make the `ActiveSessionCurrentItemUiState` nullable inside `MainContentUiState`, we can save ourselves the `visbile` property and the awkward default values

mipro98 replied:
Yes, with visibility it's always difficult because IMO visibility is part of the general UI state but actually not part of the element itself whether it is shown or not so yeah might be a solution to use the Nullable type... Although I don't agree with the default values, see above.

mipro98 replied:
I tried to make them nullable (via `?`) but then in Compose, smart-casting is not possible anymore, so is early returning when the state is `null`. That means we would always have to define unnecessary fallbacks for `null` even though we don't use the function.

Wrapping the States with `Nullable<>` makes them smart-castable in Composables but then the StateFlow `combine` functions complain about mismatched types because the compiler interprets them as `Nullable<out StateClass>>` rather than `Nullable<StateClass>` 🤔

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/7438c55b-c970-48ba-a996-75c5142c4d02)

(this also happens with `initialValue = Nullable(null)`)

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+
+@Stable
+data class ActiveSessionToolsUiState(
+    val activeTab: ActiveSessionTab = ActiveSessionTab.entries[0],
```
I like to add a companion object to classes like `ActiveSessionTab` which includes a constant value `DEFAULT`

mipro98 commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionUiState.kt:
```diff
@@ -8,20 +8,144 @@
+    val timerUiState: StateFlow<ActiveSessionTimerUiState> = MutableStateFlow(ActiveSessionTimerUiState()),
+    val currentItemUiState: StateFlow<ActiveSessionCurrentItemUiState> = MutableStateFlow(ActiveSessionCurrentItemUiState()),
+    val pastSectionsUiState: StateFlow<ActiveSessionCompletedSectionsUiState> = MutableStateFlow(ActiveSessionCompletedSectionsUiState()),
+    val endDialogUiState: StateFlow<ActiveSessionEndDialogUiState> = MutableStateFlow(ActiveSessionEndDialogUiState())
```
Having a `visible` flag has the advantage that the comments and rating a user maybe wrote in the field stays saved after the user dismisses the dialog. This way could maybe add some comments along the way and they all stay in the dialog when it is opened.

matthiasemde replied:
Touché if that is indeed the desired behavior.  🤔

matthiasemde replied:
Actually, no. Just because the UiState is null doesn't mean we can't store the comment in the ViewModel

mipro98 replied:
Yes, maybe it would be better anyways to make it properly and store it somewhere more persistent than in the ViewModel (which will be destroyed on app close)

matthiasemde replied:
In any case, the `visible` flag is kind of redundant and in the case it is `false`, all other properties values are irrelevant so the whole object might as well be `null`

matthiasemde commented:
Personal opinions on the current UI
- I think the "tools" bottom sheet should be fully collapsible (at which point the tab indicator should disappear)
- The practice timer could be higher up. The tool bottom sheets get uncomfortably close when fully extended 
- I'm not a fan of the "add item" dialog opening immediately 
- The fab should be visible when a tool is extended (don't ask me how to implement it)
- I don't think "add item" is the best text. I'd prefer something like "Start practicing" which changes to "Next item"

mipro98 commented:
>     * I think the "tools" bottom sheet should be fully collapsible (at which point the tab indicator should disappear)

Okay, yes i would be okay with that to have more space. I also thought about that but only on very small screens but discarded the idea because it would be too inconsistent. I thought about having the "tools" as a core functionality of the UI, even more "important" than the reference of "already practiced" items so it should be always visible.

How would you then define the tap behavior for the tabs? Closed -> Peek -> Open -> Peek -> Closed?  Or should tapping still only toggle between peek and open and only close it by drag gesture?

>     * The practice timer could be higher up. The tool bottom sheets get uncomfortably close when fully extended

The practice timer starts at the center and is centered together with the sections list. When sections list becomes full it is fully expanded. I think you're talking about this case?

Actually, the Practice Timer is as high as possible. Above the timer is only the [`AppBar`](https://m3.material.io/components/top-app-bar/overview) which takes some space... It's the default space occupied by the `AppBar`

>     * I'm not a fan of the "add item" dialog opening immediately

It's basically for (1) discoverability and user guidance: when it opens the user immediately knows "what to do" since he is prompted to start the session by adding the first item (so we definitely exclude that someone is confused and searches for a "start" button or so).  And (2) for practicability: basically there is no use in seeing an empty ActiveSession on start and the user will anyways press on the add item button since it's the only action. So it is much quicker and less annoying for the user.

If we don't show the popup directly, we at least should rename the FAB to "Start session" or somthing like that.

>     * The fab should be visible when a tool is extended (don't ask me how to implement it)

Where it should be? There's no real space for it. Some rough ideas:
* The bottom sheet could push it to the top until it resides anywhere above it (can be implemented with `requireOffset()` from bottomSheetState).
* The Fab could re-appear somewhere else the moment it is covered by the bottom sheet.

Aside from accessibility of the FAB, I'm also not so happy with the FAB being covered from the bottomSheet since hierarchically the FAB should be the highest layer on the screen...

>     * I don't think "add item" is the best text. I'd prefer something like "Start practicing" which changes to "Next item"

Sounds good 👍🏻 (I was not creative enough for that)

matthiasemde commented on app/src/main/java/app/musikus/usecase/activesession/GetSessionTimerState.kt:
```diff
@@ -0,0 +1,37 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+enum class SessionTimerState {
```
Do we really need a separate `SessionTimerState`? Can't we consolidate it in a single `ActiceSessionState`?

mipro98 replied:
Yes, this bothers me also a bit. I thought of the state as a kind of *domain-level* state which gives us a `Flow` to observe events like pausing, starting, etc. to consume it in the UI. The `sessionState` in `ActiveSessionViewModel` is then a *UI-state*, but [currently there is a 1:1 mapping](https://github.com/matthiasemde/musikus-android/blob/1d6f8c1c873919c9158ec292a3d9ddbaf811c1c3/app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt#L75), so kind of only a proxy.

Maybe in the future we want to add functionality which is only ui-specific and not domain-specific or so. If we now ditch the *ui state* object, only the `when` (and the enum class) would not be needed anymore, so a total gain of 10 LOC we're talking about 😅

matthiasemde replied:
As I said in my other comments I think we should differentiate at least three kinds of state: 
- _domain/data-state_ which is more or less persistent.
- _ViewModel-state_ which are StateFlows stored in and managed by the ViewModel
- _UI-state_ which is entirely derived by the previous two types of state 

Now it might be legit to have a separate `ActiveSessionState` on _domain_ and _ViewModel_ level but I'm not sure whether we need it at the moment.

mipro98 replied:
I don't see what is the difference between ViewModel and UI state. The ViewModel should contain the UI state

matthiasemde replied:
Yes, and no. In our case the ViewModel contains the logic which is used to derive the UiState from the data coming from the domain layer + the non persistent state which the ViewModel stores itself.

I concede that saying the UiState is entirely derived is kind of paradoxical since then it isn't really state anymore but still...

matthiasemde commented on build.gradle.kts:
```diff
@@ -12,3 +12,21 @@ buildscript {
         classpath("de.mannodermaus.gradle.plugins:android-junit5:1.10.0.0")
     }
 }
+subprojects {
```
Does this have to be in the top level Gradle file? Since we are building a single module application ATM it probably doesn't really matter, but in that case I'd put everything into the module gradle file.

mipro98 replied:
Oh okay, maybe we can move it then. I just followed the instructions [here](https://developer.android.com/develop/ui/compose/performance/stability/diagnose#setup)

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt:
```diff
@@ -3,416 +3,350 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2024 Michael Prommersberger, Matthias Emde
+ * Copyright (c) 2024 Michael Prommersberger
```
😥

mipro98 replied:
> To my defense, I did not remove it, but I created a completely new file and re-wrote everything from scratch so probably copy-pasted some license header... 😅

matthiasemde replied:
Fair enough then 😅

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt:
```diff
@@ -3,416 +3,350 @@
-    /**
-     *  ---------------- Composing the Ui state --------------------
-     */
+    /** ------------------- Own StateFlow UI states  ------------------------------------------- */
```
IMO it would be a good rule for us to never explicitly store _UiState_ in a variable inside the ViewModel and only derive it from private StateFlow.

mipro98 replied:
but they are *private* StateFlows, this is basically the lowest-level UI state there is, because they are not derived from UseCases. since all are packed in one StateFlow and will have to be updated together anyways, it makes no sense to define another variable for every property plus another state for wrapping all the properties inside the data class.

There would be no functional benefit but largely more useless boilerplate code, or am I wrong?

matthiasemde replied:
I still like the idea, that the ViewModel has its _own_ stateFlows, _imported_ stateFlows and maybe _combined_ stateFlows which comprise the ViewModel **state**. And from that state, the UiState is entirely derived even if it is a 1-to-1 mapping of a ViewModel state to UiState. This way we never write to UiState.

mipro98 replied:
I agree 👍. That's exactly what I did. I never used the StateFlows from the useCases but created own ones, even though they are just 1:1 mappings (that's why I called them proxies).

So all combine statements compose themselves only from private StateFlows of the ViewModel

matthiasemde replied:
I guess it's just a naming convention at that point. I would have omitted the "UiState" suffix for these StateFlows.

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt:
```diff
@@ -3,416 +3,350 @@
-            is ActiveSessionUiEvent.DiscardSessionDialogConfirmed -> {
-                discardSession()
-                _showDiscardSessionDialog.update { false }
+            ActiveSessionUiEvent.ToggleCreateFolderDialog -> _dialogVisibilities.update {
```
I don't feel like there is a need to group the dialog visibilities...

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt:
```diff
@@ -428,28 +362,7 @@ class ActiveSessionViewModel @Inject constructor(
-
     private fun stopSession() {
-        val endDialogData = _endDialogData.value ?: return
         applicationScope.launch {
```
Can you remind me again why this is inside the applicationScope? 😅

mipro98 replied:
`getFinalizedSession` is suspend
`sessionUseCases.add` is suspend

matthiasemde replied:
Yes, but why not use the ViewModel scope?

matthiasemde commented on app/src/main/java/app/musikus/usecase/activesession/GetSessionTimerState.kt:
```diff
@@ -0,0 +1,37 @@
+    UNKNOWN
+}
+
+class GetSessionTimerState (
```
~~`GetSessionTimerState`~~ -> `GetSessionTimerStateUseCase`

matthiasemde commented on app/src/main/java/app/musikus/ui/library/LibraryScreen.kt:
```diff
@@ -440,7 +446,6 @@ fun LibraryContent(
                 ) {
                     LibraryUiItem(
                         modifier = Modifier.padding(
-                            vertical = MaterialTheme.spacing.small,
```
I noticed that the paddings around the library items are pretty large. Is this intentional?

matthiasemde replied:
Ok, I figured as much. Let's review the Library screen together. Or in another PR?

mipro98 replied:
I removed the additional horizontal padding there was from before and now they look normal

matthiasemde commented on app/src/main/java/app/musikus/ui/goals/GoalsViewModel.kt:
```diff
@@ -460,4 +460,4 @@ class GoalsViewModel @Inject constructor(
             clearDialog()
         }
     }
-}
```
👀

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSession.kt:
```diff
@@ -11,104 +11,188 @@
+import java.time.ZonedDateTime
+import java.util.UUID
+import kotlin.random.Random
+import kotlin.time.Duration.Companion.seconds
```
Holy schmoly 🤯

mipro98 replied:
?

`UUID` is needed in L#1248
`ZonedDateTime` is needed in L#1432 and L#1696
`Random` is needed in various previews to generate random content
`seconds` is needed in `PreviewActiveSessionScreen`

I have auto-import activated, so all imports should actually be needed somewhere

matthiasemde replied:
It's not a critique 😅 just a comment

matthiasemde commented on app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt:
```diff
@@ -3,416 +3,350 @@
     }
 
-
+    private fun onEndDialogUiEvent(event: ActiveSessionEndDialogUiEvent) {
```
Ah I missed that. In that case: very nice 👌

matthiasemde commented on app/src/main/java/app/musikus/ui/components/SwipeToDeleteContainer.kt:
The license header should be at the top of the file wit the package name below it.

Apart from that I am absolutely in love with this Component ❤️

matthiasemde commented on app/src/main/java/app/musikus/ui/theme/Theme.kt:
~` * Copyright (c) 2024 Matthias Emde`~
` * Copyright (c) 2024 Matthias Emde, Michael Prommersberger`

matthiasemde commented on app/src/main/java/app/musikus/ui/theme/Theme.kt:
```diff
@@ -74,43 +79,142 @@ fun MusikusTheme(
+            )
+        }
     }
-
```
Why remove this empy line? I think it adds a nice separation between the conditional block above and the main content of the whole `MusikusTheme()`

matthiasemde commented on app/src/main/java/app/musikus/ui/theme/Theme.kt:
```diff
@@ -74,43 +79,142 @@ fun MusikusTheme(
+@Preview(name = "Dark", group = groupNameElem8, uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL)
+annotation class MusikusPreviewElement8
+
+
```
IMO theme properties which affect the actual app should not come after preview related code. If we can not move the preview code to separate files, we should at least keep it apart from the main code.

mipro98 replied:
I created a file `theme/PreviewStyling` which now contains all the Preview-related code

matthiasemde commented on app/src/main/java/app/musikus/ui/theme/Theme.kt:
```diff
@@ -119,13 +223,31 @@ data class Spacing(
+    /** new simple tools concept properties */
+    val toolsHeaderHeight: Dp = 95.dp,
+    val toolsBodyHeight: Dp = 205.dp,
+    /** finally the new final shit */
```
oh my

matthiasemde commented on app/src/main/java/app/musikus/ui/theme/Theme.kt:
```diff
@@ -74,43 +79,142 @@ fun MusikusTheme(
+
+/** Convenience Templates for predefined Compose Preview groups **/
+
+private const val groupNameAllLight: String = "All Elements - Light"
```
This whole block seems pretty hacky to me... Why do we need all these different annotation classes `MusikusPreviewElementX`?
And don't you think, we could store these in a separate file? 😅

mipro98 replied:
Hm you're right it's a bit "Hacky" but it's the best solution there is now for using Android Studio Preview Grouping. These values are used in the `@Preview` Annotations and have to be "hardcoded", it's not possible to use a parameter...

The purpose is to have "a bunch of Slots" for different UI elements to have previews. These Group arguments are used to group different styles of the same element together in Android Studio (dark style + color schemes + eventually different screen sizes):

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/c611dc8e-c70b-4abd-a6aa-a5b67c4c2fc7)


for example, "Element 5" is the Library Item. The `group` argument makes Android Studio group them all together so they can be displayed together: 

![image](https://github.com/matthiasemde/musikus-android/assets/28269133/8a270def-cc55-45cb-958c-cf333a5590cd|width=100px)


This is "the best I found" - with the only downside of having to remember which element is "Element X" in each screen - but you can quickly skip through. If you have a better solution for this problem, I'd be glad for ideas :)

mipro98 replied:
> And don't you think, we could store these in a separate file? 😅

I now created an extra file, `PreviewStyling.kt`

mipro98 commented:
I addressed all the comments except the one with the Grade about the compiler reports.

I noticed that the recorder is still buggy somehow. Especially the service + notification needs some optimization and the WaveForm causes MemoryException on long recordings.

I can look into the Waveform issue maybe but I fear we will maybe need a workaround for long recordings for now...

matthiasemde commented:
> I noticed that the recorder is still buggy somehow. Especially the service + notification needs some optimization and the WaveForm causes MemoryException on long recordings.
> 
> I can look into the Waveform issue maybe but I fear we will maybe need a workaround for long recordings for now...

I think we can fix the Recorder in #96 and #25.

I'll have one last look at the code and then I'll hopefully approve it :)

mipro98 commented:
> I think we can fix the Recorder in #96 and #25.

Ah yes, totally forgot about #25 😅

matthiasemde commented:
I just started my review and the thing I immediately noticed is the jumping around of the fab and the main content.

https://github.com/matthiasemde/musikus-android/assets/53527522/b5c317db-5806-467c-a428-7fce8f90c359


Maybe we should look at this in a dedicated issue.

mipro98 commented:
> I just started my review and the thing I immediately noticed is the jumping around of the fab and the main content.
> screen-20240616-122130.mp4
> 
> Maybe we should look at this in a dedicated issue.

That we already discussed a lot and was already approved by you. When I implemented this, I [already wrote](https://github.com/matthiasemde/musikus-android/pull/84#issuecomment-2081437290):
> However, this results in a bit sketchy jiggling of the UI when collapsing the layout (because of the centering and the weights).

As a reminder, this comes through the fact that `BottomSheetScaffold` is not meant to be fully collapsed[^1] but instead should *"co-exist with the main content"*.

This is the reason the returned contentPadding does not work when it's hidden and we have to hack manually changing the padding to 0dp instead.

This sucks but if we want to fully collapse the bottomSheet, we have no alternative other than re-writing our own StandardBottomSheet implementation :/

[^1]: somehow related to [this issue](https://issuetracker.google.com/issues/276929560)

mipro98 commented:
I now completely disabled the auto-hiding of the tools tab since it is, in the current state, too janky:

  * Takes some delay (for State to initialize) until it is hidden after entering ActiveSession which is irritating
  * breaks deep links since the bottom sheet is hidden after we scrolled to the tab because of the deep link

The underlying issue is that the ActiveSession has no means to query whether the Metronome and / or the Recorder is running since all this info in encapsulated inside their own ViewModels. We somehow need a clean solution for that since this will also enable us e.g. to add a red dot button next to the Recorder tab or handle starting the metronome alongside the recorder or auto-naming the recordings based on the current item...

We could inject both viewModels inside activeSession but I consider this not so nice since then all the logic described above will be handled in the UI rather than the ViewModel.


---

Also, I now included basic permission checks, especially for notification permissions since this wasn't done at all. I tried to re-use the PermissionsUseCases for that but however there is still some code repetition (extra Exception Channels for each ViewModel, ...) and missing functionality (show a Dialog to the user, offer a setting to take him to settings after 2x decline,...).

Currently, only a toast is shown on decline but at least everything is checked. This is something to be refined in another PR.

matthiasemde reviewed (approved):
LVGTM 🔥💯👏
